packages/proxy/lib/http/response-middleware.ts

```diff
@@ -331,11 +388,89 @@ const MaybePreventCaching: ResponseMiddleware = function () {
   this.next()
 }
 
+const determineIfNeedsCrossOriginHandling = (ctx: HttpMiddlewareThis<ResponseMiddlewareProps>) => {
+  const previousAUTRequestUrl = ctx.getPreviousAUTRequestUrl()
+
+  // A cookie needs cross origin handling if it's an AUT request and
+  // either the request itself is cross-origin or the origins between
+  // requests don't match, since the browser won't set them in that
+  // case and if it's secondary-origin -> primary-origin, we don't
+  // recognize the request as cross-origin
+  return (
+    !!ctx.req.isAUTFrame &&
+    (
+      (previousAUTRequestUrl && !cors.urlOriginsMatch(previousAUTRequestUrl, ctx.req.proxiedUrl))
+      || !ctx.remoteStates.isPrimaryOrigin(ctx.req.proxiedUrl)
+    )
+  )
+}
+
+interface EnsureSameSiteNoneProps {
+  cookie: string
+  browser: Browser | { family: string | null }
+  isLocalhost: boolean
+  url: URL
+}
+
+const cookieSameSiteRegex = /SameSite=(\w+)/i
+const cookieSecureRegex = /Secure/i
+const cookieSecureSemicolonRegex = /;\s*Secure/i
```

Comment:
We shouldn't use regex here. Instead, we should use a RFC-compliant cookie parsing/serializing library like https://github.com/jshttp/cookie.

Otherwise, we will mishandle plenty of cookies in confusing ways, like:

```http
Set-Cookie: insecure=true # matches cookieSecureRegex
```

Reply:
@chrisbreiding can you link to the task here plz? this seems fine for experimental

Reply:
https://miro.com/app/board/o9J_lZY-DnE=/?moveToWidget=3458764523720640348&cot=14